text/2021-12-09-logging.md

```diff
@@ -0,0 +1,302 @@
+recommend to use:
+
+- Error() - Error should be used to indicated unexpected error, for example,
+  unexpected errors returned by subroutine function calls.
+
+- Info() - Info should be used as a human journal or diary. And it could be used
+  to log expected error as warning.
+
+- Info() hash multiple levels:
+  - V(0) - This is the default level, would ALWAYS be visible to users.
+    - CLI argument handling, print args before the program starts
+    - Application configuration
+    - Expected errors as warning message
+    - Journal for application's major behavior
+      - service endpoint: resolve requests
+      - long-time batch-job: job start, job complete
+    - Acquirement on system resources: listen on port, persistent in file, etc.
+  - V(1) - Default log level for debug.
+    - Expected error that repeat frequently, relates to conditions would be
+      corrected, like `StatusReasonConflict` when reconciling
+    - Component configuration and initialize
+  - V(2) - Useful state changes and conditional branch.
+    - State which is useful for debugging changed
+    - Choose a branch in a conditional statement
+  - V(3) - Extended information about changes
+    - Logging with state updates like `err := updateState(); err == nil`
+    - Logging error with more context before wrap and return it
+  - V(4) - Debug level verbosity, any other behavior that changes the state
+  - V(5) - Default log level for trace.
+    - Progressing in a for-loop
+    - Omitted things within "best effort" pattern
+    - Use default value with not configured value
+  - V(6) - Communication with other components
+    - Logging in handler before processing/resolving request
+    - RPC call with other components/services
+  - V(7) - Detailed information about communication
+    - Detailed payload and status of request/response within RPC
+
+The number of level still limited by the backend logging framework, for example,
+you could only create 128 level in zap.
+
+### Relations between error and logging
+
+You could resolve an error by log it with `Error()`, it usually means this
+program could not handle this error properly, and let the user know. But if your
+codes could handle this error, you should not log it with `Error()`.
+
+You could also use debug level `Info()` to log error, with more detailed context
+and behaviors.
+
+## Drawbacks
+
+<!-- Why should we not do this? -->
+
+## Alternatives
+
+<!-- - Why is this design the best in the space of possible designs?
+- What other designs have been considered and what is the rationale for not
+  choosing them?
+- What is the impact of not doing this? -->
+
+### Why zap? And why not other logger framework as logr backend
+
+There are still many other logging frameworks supported by logr:
+https://github.com/go-logr/logr#implementations-non-exhaustive. I select zap
+only because I am familiar with it, and I think the API of zap is easy to use
```

Comment:
maybe we could use the same log in TiDB

Reply:
If we decide to use `logr`, we could build a bridge easily between pingcap/log and logr.

And I prefer to use "raw" zap rather than use pingcap/log, the latter one does not provide many useful features because we would not persist log files and rotation log files. 🤔